### PR TITLE
[FLINK-13540] DDL do not support properties key that contains number …

### PR DESCRIPTION
**Pull request by danny0405**
## What is the purpose of the change

This patch changes the table option keys format to string literal, they are as sql identifiers before.

## Brief change log

  - Change the table option key to string literal

## Verifying this change

See tests in FlinkSqlParserImplTest.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`:  no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl [hunk 1]
```diff
@@ -115,18 +115,18 @@ void UniqueKey(List<SqlNodeList> list) :
     }
 }
 
-SqlNode PropertyValue() :
+SqlNode TableOption() :
 {
-    SqlIdentifier key;
+    SqlNode key;
```
Comment (wuchong) [inline flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl]: Force key and value to be `StringLiteral`?

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/TableOption.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Table options of a DDL, a key-value pair
+ * with both key and value as string literal.
+ */
+public class TableOption extends SqlCall {
```
Comment (wuchong) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/TableOption.java]: Rename it to `SqlTableOption` or `SqlTableProperty` and move to `org.apache.flink.sql.parser.ddl`?

File: flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java [hunk 3]
```diff
@@ -454,12 +454,36 @@ public void testCreateInvalidPartitionedTable() {
...
+			"  'a-b-c-d124' = 'ab',\n" +
+			"  'a.b-c-connector.e-f.g' = 'ada',\n" +
+			"  'a.b-c-d.e-1231.g' = 'ada',\n" +
+			"  'a.b-c-d.*' = 'adad')\n";
+		String expected = "CREATE TABLE `SOURCE_TABLE` (\n" +
+			"  `A`  INTEGER,\n" +
+			"  `B`  BIGINT,\n" +
+			"  `C`  STRING\n" +
+			") WITH (\n" +
+			"  'a-b-c-d124' = 'ab',\n" +
```
Comment (wuchong) [inline flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java]: nit: could you add a key `a.b.1.c` ?

File: flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl (hunk 1 above)
Comment (danny0405) [reply flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl]: The `StringLiteral()`'s return type is `SalNode`, i suggest to keep the type as it is(do not force type cast in parser), we have a type cast in `TableOption`.

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/TableOption.java (hunk 2 above)
Comment (danny0405) [reply flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/TableOption.java]: Agree, done

File: flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java (hunk 3 above)
Comment (danny0405) [reply flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java]: Added

**Comment by danny0405** [comment]
@wuchong  Thanks for the review, i would rather to keep the `SqlProperty`, because it is a common pattern, we may still need to use this structure for future grammars, i.g. `CLUSTER BY` .

**Comment by twalthr** [comment]
+1 to this from an API perspective.

**Pull request closed without merging**